feat(battery): add option to show charge percentage next to the icon

ilya-zlobintsev opened the pull request:
Closes #735 

<img width="368" height="579" alt="image" src="https://github.com/user-attachments/assets/a190628e-631c-4f2a-b4a2-62e64e5e52c3" />

I haven't tested how the applet looks when the GPU selector is visible, as I don't currently have a hybrid graphics system, but it should be fine.

ilya-zlobintsev commented:
I've discovered that this causes the panel to resize when it's in vertical mode, it's probably not something that should be happening due to a widget, will try to check it out later

al12gamer commented:
Looking forward to this

Cheong-Lau commented on cosmic-applet-battery/src/app.rs:
```diff
@@ -509,13 +567,14 @@ impl cosmic::Application for CosmicBatteryApplet {
-                    .into(),
-                PanelAnchor::Top | PanelAnchor::Bottom => Row::with_children(vec![btn, dot])
+            if is_horizontal {
+                Row::with_children(vec![btn, dot])
```
🤓 if `Row::with_children` and `Column::with_children` takes an `impl IntoIterator`, then the `vec!` macro isn't really needed, you could just pass the array itself

... 😅 sorry

git-f0x replied:
I think it would be better to use `with_children` (since that's used basically everywhere applicable). Just remove the `vec!`.

dyrland commented:
Looking forward to this!

mmstick commented:
@maria-komarova Would we want to show a toggle for battery percentage in the battery applet?

ilya-zlobintsev commented:
>  @ilya-zlobintsev Could you remove the toggle from the applet's view? We will put this in cosmic-settings instead.

Done. 

Is there anything needed for the applet's config schema to be usable from cosmic-settings? I am not sure how the settings are shared between projects.

mmstick commented:
@ilya-zlobintsev I checked and you should put the config into the cosmic-applets-config crate. The time applet is here: https://github.com/pop-os/cosmic-applets/blob/master/cosmic-applets-config/src/time.rs. cosmic-settings imports this crate. Technically you could manually request a config from a namespace with cosmic-config directly, but having a config crate helps with ensuring that the config has a single source of truth.

ilya-zlobintsev commented:
I've moved the config struct to `cosmic-applets-config` and also added a subscription for config changes. It works now when manually editing the config file. Though I'm not sure about the current app name `CosmicAppletButton`, maybe it should be changed to `CosmicAppletBattery`?

wash2 commented:
Be sure to test this with the applet forced to the overflow menu.

git-f0x commented:
I think the overflow can only display square items, since e.g. the audio applet media controls don't show up when it's in the overflow. So maybe this is fine for now?

leviport reviewed (changes requested):
There are a couple changes I'd like to see.

1. This was discussed a bit already, but the config file needs to be added in `~/.config/cosmic/com.system76.CosmicAppletButton/v1` to make this work, and that name made it difficult for me to find at first. I think it would be a lot better if we could change `CosmicAppletButton` to `CosmicAppletBattery`. This is in line with other applet configs (`CosmicAppletTime`, `CosmicAppletAudio`, `CosmicAppletInputSources`, etc.) and it is more descriptive.

2. I noticed that when the battery is at 100%, the right edge of the `%` symbol is clipped off slightly:
<img width="557" height="78" alt="Image" src="https://github.com/user-attachments/assets/ce6d1391-ef47-4dc9-acb6-40b7bc616470" />

Can we make sure that doesn't get clipped off like that?

Other than these nitpicks, this is working nicely. Thanks for this contribution!

RafaelKr commented:
I just updated my system and also did a restart, but I don't see that option (yet?):
<img width="360" height="514" alt="image" src="https://github.com/user-attachments/assets/6bd18c69-565e-4fcc-aa14-84b4c8819c21" />

Is the update maybe not available via apt yet?
```
dpkg --list | grep cosmic-applets
ii  cosmic-applets                          1.0.2~1770069480~24.04~57e55a9
```

Also I don't see that `80% maximum charge` toggle, is this only available to some devices? On my Lenovo device it's also possible to set the so-called conservation_mode via `echo 1 > /sys/bus/platform/drivers/ideapad_acpi/VPC2004\:00/conservation_mode`, are there any plans to support this?

maxdec commented:
@RafaelKr if you want to enable it right now, you can do this:

```
mkdir -p ~/.config/cosmic/com.system76.CosmicAppletBattery/v1
echo "true" > ~/.config/cosmic/com.system76.CosmicAppletBattery/v1/show_percentage
```

(or something close... I'm not on my PopOS laptop at the moment)

then reboot.
